First checkin of frontend

Wolfsblvt opened the pull request:
Checkin of buttons + js animation and also start of ajax handling.

NOTE: For testing purposes, the callback of ajax will be executed every
time. This needs to be adjusted when server side ajax handling is
implemented.

Wolfsblvt committed:
First checkin of frontend

Checkin of buttons + js animation and also start of ajax handling.

NOTE: For testing purposes, the callback of ajax will be executed every
time. This needs to be adjusted when server side ajax handling is
implemented.

Wolfsblvt commented:
Yeah, I have used most of your code @VSEphpbb, just needed a few small modifications so that callback function will be executed.

iMattPro commented on styles/all/theme/collapsiblecategories.css:
```diff
@@ -0,0 +1,49 @@
+.hidden-category + .forabg .topiclist.forums {display: none}
```
Lets add a comment above this line explaining what this is for/doing

iMattPro replied:
Also lets be consistent with the rest of our css, so `{ display: none; }`

iMattPro commented on styles/all/theme/collapsiblecategories.css:
```diff
@@ -0,0 +1,49 @@
+	top: 3px;
+	width: 0;
+}
+a.collapse-hide:before { border: none; }
```
missing newline at end of file

iMattPro commented on styles/all/template/js/collapsiblecategories.js:
```diff
@@ -0,0 +1,58 @@
+
+	'use strict';
+
+	$(function () {
```
We do not need this. This is shorthand for document ready right? We do not need to document ready since this script is loaded in the footer, so it is loaded after the DOM above already exists.

iMattPro commented on styles/all/template/js/collapsiblecategories.js:
```diff
@@ -0,0 +1,58 @@
+			}
+
+			// Add button
+			$header.append('<a class="collapse-btn collapse-' + ((hidden) ? 'show' : 'hide') + '" data-id="' + id + '" data-ajax="phpbb_collapse"></a>');
```
An alternative to hardcoding all this HTML, which I think would be better, and more readable, is this:

``` javascript
var $button = $('<a>')
    .addClass('collapse-btn collapse-' + ((hidden) ? 'show' : 'hide'))
    .attr({
        href: collapsible_categories_url + id,
        'data-ajax': 'phpbb_collapse',
        'data-overlay': true
    });
$header.append($button);
```

Also we need an href for the data-ajax to work. The var `collapsible_categories_url` would be set back in a template event like:

```
<script>
    var collapsible_categories_url = '{{ ROOT_PATH }}collapse/';
</script>
```

Note that I'm using TWIG syntax there :)

Wolfsblvt replied:
As I have thought about this now, why root_path? Can whe expect that controller routes are rewritten? And why without sid?
Or was it just a suggestion that we have to do something there?

I mean I would have done 

``` php
$this->helper->route('phpbb_collapsible');
```

and put that in a template var.

iMattPro replied:
Yeah you can get rid of that script tag... we should define it once we have set up our routings

iMattPro commented on styles/all/template/js/collapsiblecategories.js:
```diff
@@ -0,0 +1,58 @@
+			}
+		});
+
+		// Since we dynamically create our AJAX button, phpBB's does not ajaxify it
```
oops, should be phpBB not phpBB's

iMattPro commented on styles/all/template/js/collapsiblecategories.js:
```diff
@@ -0,0 +1,58 @@
+
+		// Since we dynamically create our AJAX button, phpBB's does not ajaxify it
+		// so we reproduce phpBB's ajaxify on our buttons here
+		$('.collapse-btn').each(function () {
```
Very interesting... Because you put the INCLUDEJS in overall_header, our script gets loaded before phpBB's ajax.js. Which means we do not need this function (lines 26~40). Right now our button is being double-ajaxified, so we should remove this code :)

Wolfsblvt replied:
I am so sure I have tested that cause I thought the same, but it didn't work.
Now I can confirm what you say. Maybe cause of my document.ready() wich put it behind the ajax.js?
I will remove that now.

iMattPro commented on styles/all/template/js/collapsiblecategories.js:
```diff
@@ -0,0 +1,58 @@
+		phpbb.addAjaxCallback('phpbb_collapse', function (res) {
+			var $this = $(this),
+				forum = $this.attr('data-id');
+			if ('just for test of animation') { // TODO: Remove this line can use the one below if ajax server side is working
```
We shouldn't commit test code like this. Since there is no PHP backend yet, we can assume the AJAX part is not functional yet. Usually just a comment in the PR description about such limitations are better. :)

Wolfsblvt committed:
Update frontend to match suggestions

Removed unused code and added some comments.
Also jQuery adding of the button is changed for better readability.

Wolfsblvt commented:
Okay, I have worked on everything you have suggested.

For now, clicking on the buttons will show an "Not found" error, cause ajax on server side does not exist.
So no debug code there anymore :P

iMattPro commented on styles/all/template/event/forumlist_body_category_header_before.html:
```diff
@@ -0,0 +1,7 @@
+<!-- IF forumrow.S_IS_CAT -->
+<a class="category <!-- IF forumrow.S_FORUM_HIDDEN -->hidden-category<!-- ENDIF -->"
```
We should move the space after `category` to inside the conditional before `hidden` so that it renders as `<a class="category">` when there is no hidden category

iMattPro replied:
We should also indent this 1 level since it's inside an <!-- IF --> block

iMattPro commented on styles/all/template/event/forumlist_body_category_header_before.html:
```diff
@@ -0,0 +1,7 @@
+<!-- IF forumrow.S_IS_CAT -->
+<a class="category <!-- IF forumrow.S_FORUM_HIDDEN -->hidden-category<!-- ENDIF -->"
+	id="category{forumrow.FORUM_ID}"
+	data-id="{forumrow.FORUM_ID}"
```
we should make this `fid_{forumrow.FORUM_ID}` if we want to use fid_1, fid_2 etc for our ids

Wolfsblvt committed:
Update work of ajax path + data-id

Wolfsblvt commented:
Updated the latest suggestions here.

PayBas commented on styles/all/template/event/forumlist_body_category_header_before.html:
```diff
@@ -0,0 +1,7 @@
+<!-- IF forumrow.S_IS_CAT -->
```
If you are in viewforum.php (with subforums), there might not be a category for some of these forums, but you could still want to hide the subforums of the particular forum.

Example: https://www.phpbb.com/community/viewforum.php?f=446

So you might want to be able to collapse "uncategorized" forums as well using `<!-- IF forumrow.S_IS_CAT or forumrow.S_FIRST_ROW or forumrow.S_NO_CAT  -->`

The forum ID would simply be the ID of the first uncategorized forum.

Something to think about.

iMattPro replied:
From what I can see, the current implementation appears to work with subforums in viewforum.php

iMattPro replied:
Oh I see what you mean... "uncategorized"... yeah, something to think about ;)

PayBas commented on styles/all/template/js/collapsiblecategories.js:
```diff
@@ -0,0 +1,46 @@
+	$('a.category[data-id] + .forabg').each(function () {
+		var $this = $(this),
+			id = $this.prev().attr('data-id'),
+			hidden = $this.prev().attr('data-hidden'),
```
Using `prev()` could potentially cause problems with other extensions that insert elements. Better use a find/child (and reuse it rather than finding it twice).

edit: wait, the el is outside the `.forabg`... hmmm... there's still gotta be a better way

edit 2: If you move the code from `forumlist_body_category_header_before.html` to `forumlist_body_category_header_after` it will be _inside_ the `.forabg` div, so you wont have to use `prev()`

iMattPro replied:
Do you have time to maybe fix this in a quick PR?
